jerry-core/mem/mem-poolman.cpp

```diff
@@ -86,61 +165,325 @@ mem_pools_init (void)
+  {
+    next_free_chunk_p = free_chunk_iter_p->u.free.next_p;
+
+    VALGRIND_NOACCESS_SPACE (free_chunk_iter_p, MEM_POOL_CHUNK_SIZE);
+  }
+} /* mem_pools_collect_empty_pools */
+
 /**
  * Long path for mem_pools_alloc
- *
- * @return true - if there is a free chunk in mem_pools,
- *         false - otherwise (not enough memory).
  */
-static bool __attr_noinline___
+static void __attr_noinline___
 mem_pools_alloc_longpath (void)
 {
-  /**
-   * If there are no free chunks, allocate new pool.
-   */
-  if (mem_free_chunks_number == 0)
+  mem_check_pools ();
+
+  JERRY_ASSERT (mem_free_chunk_p == NULL);
+
+  JERRY_ASSERT (MEM_POOL_SIZE <= mem_heap_get_chunked_block_data_size ());
+  JERRY_ASSERT (MEM_POOL_CHUNKS_NUMBER >= 1);
+
+  mem_pool_chunk_t *pool_start_p = (mem_pool_chunk_t*) mem_heap_alloc_chunked_block (MEM_HEAP_ALLOC_LONG_TERM);
+
+  if (mem_free_chunk_p != NULL)
   {
-    mem_pool_state_t *pool_state = (mem_pool_state_t*) mem_heap_alloc_block (MEM_POOL_SIZE, MEM_HEAP_ALLOC_LONG_TERM);
+    /* some chunks were freed due to GC invoked by heap allocator */
+    mem_heap_free_block (pool_start_p);
 
-    JERRY_ASSERT (pool_state != NULL);
+    return;
+  }
 
-    mem_pool_init (pool_state, MEM_POOL_SIZE);
+#ifndef JERRY_NDEBUG
+  mem_free_chunks_number += MEM_POOL_CHUNKS_NUMBER;
+#endif /* !JERRY_NDEBUG */
 
-    MEM_CP_SET_POINTER (pool_state->next_pool_cp, mem_pools);
+  JERRY_STATIC_ASSERT (MEM_POOL_CHUNK_SIZE % MEM_ALIGNMENT == 0);
+  JERRY_STATIC_ASSERT (sizeof (mem_pool_chunk_t) == MEM_POOL_CHUNK_SIZE);
+  JERRY_STATIC_ASSERT (sizeof (mem_pool_chunk_index_t) <= MEM_POOL_CHUNK_SIZE);
+  JERRY_ASSERT ((mem_pool_chunk_index_t) MEM_POOL_CHUNKS_NUMBER == MEM_POOL_CHUNKS_NUMBER);
+  JERRY_ASSERT (MEM_POOL_SIZE == MEM_POOL_CHUNKS_NUMBER * MEM_POOL_CHUNK_SIZE);
 
-    mem_pools = pool_state;
+  JERRY_ASSERT (((uintptr_t) pool_start_p) % MEM_ALIGNMENT == 0);
 
-    mem_free_chunks_number += MEM_POOL_CHUNKS_NUMBER;
+  mem_pool_chunk_t *prev_free_chunk_p = NULL;
 
-    MEM_POOLS_STAT_ALLOC_POOL ();
-  }
-  else
+  for (mem_pool_chunk_index_t chunk_index = 0;
+       chunk_index < MEM_POOL_CHUNKS_NUMBER;
+       chunk_index++)
   {
-    /**
-     * There is definitely at least one pool of specified type with at least one free chunk.
-     *
-     * Search for the pool.
-     */
-    mem_pool_state_t *pool_state = mem_pools, *prev_pool_state_p = NULL;
+    mem_pool_chunk_t *chunk_p = pool_start_p + chunk_index;
 
-    while (pool_state->first_free_chunk == MEM_POOL_CHUNKS_NUMBER)
+    if (prev_free_chunk_p != NULL)
     {
-      prev_pool_state_p = pool_state;
-      pool_state = MEM_CP_GET_NON_NULL_POINTER (mem_pool_state_t, pool_state->next_pool_cp);
+      prev_free_chunk_p->u.free.next_p = chunk_p;
     }
 
-    JERRY_ASSERT (prev_pool_state_p != NULL && pool_state != mem_pools);
+    prev_free_chunk_p = chunk_p;
+  }
+
+  prev_free_chunk_p->u.free.next_p = NULL;
+
+#ifdef JERRY_VALGRIND
+  for (mem_pool_chunk_index_t chunk_index = 0;
```

Comment:
`for (mem_pool_chunk_index_t chunk_index = 0; chunk_index < MEM_POOL_CHUNKS_NUMBER; chunk_index++)`